### PR TITLE
 option to hide portlet when nothing to select in filter

### PR DESCRIPTION
**Pull request by djay**

### CODE REVIEW

**Comment by djay** [comment]
@petschki useful feature? We have a usecase for it but not sure its worth fixing it so it doesn't disappear in the case you are narrowing down

**Comment by JeffersonBledsoe** [comment]
Just been trying to debug why the "hide when no options" test fails on Travis but not locally. It appears the issue stems from Travis using XVFB to run the test in a headless environment, causing a timing issue and portlet setting to “Hide if empty” not being saved.
The `Wait until page contains element` xpath at the end of `Set portlet` always returned true as the portlet existed behind the modal. Changing this to wait until the modal is no longer visible waits long enough to fix the test failure

File: src/collective/collectionfilter/baseviews.py [hunk 1]
```diff
@@ -2,11 +2,13 @@
...
 
 from plone import api
+from plone.memoize.view import memoize
 
 from collective.collectionfilter import PLONE_VERSION
 from collective.collectionfilter.filteritems import get_filter_items
+from collective.collectionfilter.interfaces import IGroupByCriteria
 from collective.collectionfilter.query import make_query
-from collective.collectionfilter.utils import base_query
+from collective.collectionfilter.utils import base_query, safe_iterable
```
Comment (petschki) [inline src/collective/collectionfilter/baseviews.py]: seems that something's wrong with `bin/code-analysis` ... this is not pep8 conform

File: src/collective/collectionfilter/baseviews.py [hunk 2]
```diff
@@ -19,7 +21,7 @@
 from plone.i18n.normalizer.interfaces import IIDNormalizer
 from Products.CMFPlone.utils import safe_unicode
 from six.moves.urllib.parse import urlencode
-from zope.component import queryUtility
+from zope.component import queryUtility, getUtility
```
Comment (petschki) [inline src/collective/collectionfilter/baseviews.py]: see above

File: src/collective/collectionfilter/baseviews.py [hunk 3]
```diff
@@ -118,6 +120,7 @@ def input_type(self):
         else:
             return 'checkbox'
 
+    @memoize  # because we need to call this for is_available below
```
Comment (petschki) [inline src/collective/collectionfilter/baseviews.py]: might be huge in memory?

File: src/collective/collectionfilter/baseviews.py (hunk 1 above)
Comment (djay) [reply src/collective/collectionfilter/baseviews.py]: It's running and passing. E401 isn't in the ignore list.

File: src/collective/collectionfilter/baseviews.py (hunk 3 above)
Comment (djay) [reply src/collective/collectionfilter/baseviews.py]: It's only during the request and its only the size of the number of elements in the filter, not the whole result set. Calling it result everywhere is perhaps confusing.

**Comment by djay** [comment]
@petschki  Just to be clear the portlet is only hidden in the case where nothing has been filtered so far and it only has 1 or less options. In the case where you have selected something and used 'narrow-down" mode then it will still display. Perhaps this could be made clearer in the description.

One caveat to this that currently portlets don't respect the narrow-down mode with regard to other filter portlets which I'd count as a seperate bug. Either you want the filters to act independently or you don't. So if hide_if_empty is on and narrow-down is off, then if you filter using another portlet then your portlet could disappear.

My concern with the UI is that there is starting to be too many checkboxes and options and this one is not so easy to explain. It does however serve a useful purpose in that sometimes user permissions and other things can determine the options available and if you can't choose then its more confusing than being presented with a non-choice.

File: src/collective/collectionfilter/baseviews.py (hunk 3 above)
Comment (djay) [reply src/collective/collectionfilter/baseviews.py]: @petschki I am using the right api right? It should only last until the view is gone, ie end of request

Comment (petschki) [reply src/collective/collectionfilter/baseviews.py]: sorry for not responding earlier ... I do not know enough of the memoize internals how and when this gets invalidated ...

**Comment by petschki** [comment]
aha ... so as you mentioned this is more like a rare edge case and hard to explain. Would it be better to make this a configuration registry entry which could be set by policy packages instead of providing another checkbox to the user? just thinking loudly ...

**Comment by djay** [comment]
It needs to be set per portlet. 

It’s not that rare or hard to explain. Do you want to show the portlet when the user has no options to pick.

**Comment by djay** [comment]
If it helps the use case I need it for is workflow state. The public can only see public content so it makes so sense to give them a filter. Extranet users want to be able to filter by Internal or Public.

File: src/collective/collectionfilter/baseviews.py (hunk 1 above)
Comment (petschki) [reply src/collective/collectionfilter/baseviews.py]: you're right ... pep8 says this is ok https://pep8.org/#imports

**Comment by JeffersonBledsoe** [comment]
@petschki Tests are all passing, can we get this merged?

**Pull request merged**